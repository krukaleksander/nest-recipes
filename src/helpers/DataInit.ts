import { Inject, Injectable, OnApplicationBootstrap } from '@nestjs/common';
import { Repository } from 'typeorm';
import { Recipe } from '../database/entities/Recipe.entity';
import { MockDB } from './MockDB';
import { IRecipe } from '../interfaces';
@Injectable()
export class DataInitService implements OnApplicationBootstrap {
  constructor(
    @Inject('RECIPE_REPOSITORY')
    private recipeRepository: Repository<Recipe>,
  ) {}
  async onApplicationBootstrap() {
    console.log('app is starting');
    try {
      await Promise.all(
        MockDB.map(
          async (recipe: IRecipe) => await this.recipeRepository.save(recipe),
        ),
      );
      await this.recipeRepository.save(MockDB[0]);
    } catch (error) {
      throw error;
    }
  }
}
