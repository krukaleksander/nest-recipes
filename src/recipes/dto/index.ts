export * from './RecipeDto';
export * from './IngredientDto';
