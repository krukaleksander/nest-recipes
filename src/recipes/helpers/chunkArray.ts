import { IRecipe } from '../../interfaces';

export function chunkArray(arr: IRecipe[], chunkSize: number) {
  return arr.reduce((resultArray, item, index) => {
    const chunkIndex = Math.floor(index / chunkSize);

    if (!resultArray[chunkIndex]) {
      resultArray[chunkIndex] = [];
    }
    resultArray[chunkIndex].push(item);
    return resultArray;
  }, []);
}
