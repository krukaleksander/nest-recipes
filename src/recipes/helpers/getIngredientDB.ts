import { Repository } from 'typeorm';
import { Ingredient } from '../../database/entities/Ingredient.entity';

export function getIngredientDB(repository: Repository<Ingredient>) {
  try {
    return Promise.resolve(repository.find());
  } catch (error) {
    throw error;
  }
}
