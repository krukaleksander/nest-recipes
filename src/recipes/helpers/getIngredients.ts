import { Ingredient, Recipe } from '../../interfaces';

export function getIngredients(MockDB: Recipe[], key: string) {
  const db = MockDB;
  const arrOfAllIngredients: string[] = [];
  db.forEach((recipe: Recipe) => {
    const { ingredients } = recipe;
    ingredients.forEach((ingredient: Ingredient) =>
      arrOfAllIngredients.push(ingredient[key]),
    );
  });
  return [...new Set(arrOfAllIngredients.flat(1))];
}
