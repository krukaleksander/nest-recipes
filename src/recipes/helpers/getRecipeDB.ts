import { Repository } from 'typeorm';
import { Recipe } from '../../database/entities/Recipe.entity';

export function getRecipeDB(repository: Repository<Recipe>, id?: number) {
  try {
    return Promise.resolve(
      repository.find({
        relations: ['ingredients'],
        where: { id },
      }),
    );
  } catch (error) {
    throw error;
  }
}
