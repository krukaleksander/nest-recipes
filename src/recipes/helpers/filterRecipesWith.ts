import { IngredientDto, RecipeDto } from '../dto';

export function filterRecipesWith(products: string[], recipes: RecipeDto[]) {
  let result = recipes;
  products.forEach(
    (product: string) =>
      (result = result.filter((recipe: RecipeDto) => {
        if (
          recipe.ingredients.filter(
            (ingredient: IngredientDto) => ingredient.name === product
          ).length > 0
        )
          return true;
      }))
  );
  return Promise.resolve(result);
}