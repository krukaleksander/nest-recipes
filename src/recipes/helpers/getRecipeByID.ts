import { HttpException, HttpStatus } from '@nestjs/common';
import { RecipeDto } from '../dto';

export function getRecipeByID(
  recipes: RecipeDto[],
  id: number,
): RecipeDto[] | HttpException {
  if (!id || recipes.length < 1)
    throw new HttpException('No content', HttpStatus.NO_CONTENT);
  return recipes;
}
