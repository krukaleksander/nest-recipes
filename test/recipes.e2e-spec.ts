import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { RecipesModule } from '../src/recipes/recipes.module';
import * as superagent from 'superagent';
import { MockDB } from '../src/helpers/MockDB';

describe('Recipes (e2e)', () => {
  let app: INestApplication;
  let getFromServer: (
    endpoint: string,
    body?: Record<string, unknown>,
  ) => Promise<superagent.Response>;
  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [RecipesModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    await app.init();
  });
  beforeAll(function createCustomMethods() {
    getFromServer = async (endpoint, body?) => {
      return request(app.getHttpServer()).get(endpoint).send(body);
    };
  });
  describe('/recipes/ingredients (GET)', () => {
    const endpoint = '/recipes/ingredients';
    it('should return status 200', () => {
      return request(app.getHttpServer()).get(endpoint).expect(200);
    });
    it('should return array of strings', async () => {
      const { body: response } = await getFromServer(endpoint);
      expect(response).toEqual(expect.arrayContaining([expect.any(String)]));
    });
    it('should return array of unique strings', async () => {
      const { body: response } = await getFromServer(endpoint);
      expect(new Set(response).size === response.length).toBe(true);
    });
  });
  describe('/recipes/ingredients/types (GET)', () => {
    const endpoint = '/recipes/ingredients/types';
    const ingredientsTypes = [
      'Meat',
      'Baking',
      'Condiments',
      'Drinks',
      'Produce',
      'Misc',
      'Dairy',
    ];
    it('should return status 200', () => {
      return request(app.getHttpServer()).get(endpoint).expect(200);
    });
    it('should return array of strings', async () => {
      const { body: response } = await getFromServer(endpoint);
      expect(response).toEqual(expect.arrayContaining([expect.any(String)]));
    });
    it('should return array of unique strings', async () => {
      const { body: response } = await getFromServer(endpoint);
      expect(new Set(response).size === response.length).toBe(true);
    });
    it('should match array of ingredient types', async () => {
      const { body: response } = await getFromServer(endpoint);
      expect(response).toEqual(ingredientsTypes);
    });
  });
  describe('/recipes (GET)', () => {
    const endpoint = '/recipes';
    it('should return status 200', () => {
      return request(app.getHttpServer()).get(endpoint).expect(200);
    });
    it('should return 3 recipes if limit is set to 3', async () => {
      const { body: response } = await getFromServer(`${endpoint}?limit=3`);
      expect(response).toHaveLength(3);
    });
    it('should return limit out of range if limit is bigger than database length', async () => {
      const response = await getFromServer(`${endpoint}?limit=100`);
      expect(response.body).toEqual({
        statusCode: 400,
        message: 'limit out of range',
      });
    });
    it('should throw page out of range if there are no more pages in current limit', async () => {
      const response = await getFromServer(`${endpoint}?limit=9&page=2`);
      expect(response.body).toEqual({
        statusCode: 400,
        message: 'page out of range',
      });
    });
    it('should return elements at index 4 and 5 if limit 2 and page 3', async () => {
      const response = await getFromServer(`${endpoint}?limit=2&page=3`);
      expect(response.body).toEqual(MockDB.slice(4, 6));
    });
    it('should return elements at index 4, 5, 6 and 7 if limit 4 and page 2', async () => {
      const response = await getFromServer(`${endpoint}?limit=4&page=2`);
      expect(response.body).toEqual(MockDB.slice(4, 8));
    });
  });
  describe('/recipes/time (GET)', () => {
    const endpoint = '/recipes/time';
    it('should return status 200', () => {
      return request(app.getHttpServer()).get(endpoint).expect(200);
    });
    it('should return  empty array if 0 time is passed', () => {
      return request(app.getHttpServer())
        .get(endpoint)
        .send({ time: 0 })
        .expect([]);
    });
    it('should return one recipe if time is 6', async () => {
      const { body: response } = await getFromServer(endpoint, { time: 6 });
      expect(response).toHaveLength(1);
    });
  });
});
