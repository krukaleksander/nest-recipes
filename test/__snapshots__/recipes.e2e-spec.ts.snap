// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipes (e2e) /recipes/byproduct (GET) should return recipes with given single product 1`] = `
Array [
  Object {
    "id": 8,
    "imageURL": "http://img.sndimg.com/food/image/upload/w_266/v1/img/recipes/50/84/7/picMcSyVd.jpg",
    "ingredients": Array [
      Object {
        "id": 56,
        "name": "olive oil",
        "quantity": "1 1/2 tbsp",
        "type": "Condiments",
      },
    ],
    "name": "Roasted Asparagus",
    "originalURL": "http://www.food.com/recipe/roasted-asparagus-50847",
    "steps": Array [
      "Preheat oven to 425Â°F.",
      "Cut off the woody bottom part of the asparagus spears and discard.",
      "With a vegetable peeler, peel off the skin on the bottom 2-3 inches of the spears (this keeps the asparagus from being all.\\",string.\\", and if you eat asparagus you know what I mean by that).",
      "Place asparagus on foil-lined baking sheet and drizzle with olive oil.",
      "Sprinkle with salt.",
      "With your hands, roll the asparagus around until they are evenly coated with oil and salt.",
      "Roast for 10-15 minutes, depending on the thickness of your stalks and how tender you like them.",
      "They should be tender when pierced with the tip of a knife.",
      "The tips of the spears will get very brown but watch them to prevent burning.",
      "They are great plain, but sometimes I serve them with a light vinaigrette if we need something acidic to balance out our meal.",
    ],
    "timers": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      10,
      0,
      0,
      0,
    ],
  },
]
`;
